src/statsd-parsers/basic/basic.c

```diff
@@ -136,14 +185,40 @@ statsd_datagram* basic_parser_parse(char* buffer) {
         }
         current_segment_length++;
     }
-    if (any_tags == 1) {
-        int json_len = strlen(tags_json_buffer);
-        tags_json_buffer[json_len - 1] = '}';
-        datagram->tags = (char *) malloc(json_len + 1);
-        ALLOC_CHECK("Not enough memory to save tags JSON.");
-        memcpy(datagram->tags, tags_json_buffer, json_len);
-        datagram->tags[json_len] = '\0';
+    free(segment);
+    if ((required_fields_flags & field_allocated_flags) != required_fields_flags)  {
+        goto error_clean_up_end;
     }
+    return 1;
+
+    error_clean_up:
+    free(attr);
     free(segment);
-    return datagram;
+
+    error_clean_up_end:
+    if (field_allocated_flags & 1) {
+        free_datagram_tags((*datagram)->tags);
+    }
+    if (field_allocated_flags & 1 << 1) {
+        free((*datagram)->metric);
+    }
+    if (field_allocated_flags & 1 << 2) {
+        free((*datagram)->type);
+    }
+    if (field_allocated_flags & 1 << 3) {
+        free((*datagram)->value);
+    }
+    if (field_allocated_flags & 1 << 4) {
+        free((*datagram)->instance);
+    }
+    if (field_allocated_flags & 1 << 5) {
+        free((*datagram)->sampling);
+    }
+    if (tag_allocated_flags & 1) {
+        free(tag_key);
+    }
+    if (tag_allocated_flags & 2) {
+        free(tag_value);
+    }
+    return 0;
```

Comment:
Can't you reuse the `free_datagram` function, give it a bitarray as an (optional) parameter to selectively free fields. I am not good in C, maybe @natoscott have an advice here. Overall this looks fragile, but I am doing programming in garbage collected languages for too many years how ;-)

Reply:
I wasnt doing any valgrind correction/opmization yet either.

Reply:
Reusing the free function is still a good idea I think.

Reply:
> Can't you reuse the `free_datagram` function, give it a bitarray as an (optional) parameter to selectively free fields. I am not good in C, maybe @natoscott have an advice here. Overall this looks fragile, but I am doing programming in garbage collected languages for too many years how ;-)
> 
> Frankly, I haven't studied this part in fine detail, valgrind would probably do the job. I expect changes in this file so probably this is not wroth huge effort line by line for now.

Seems OK, although I would probably use a pre-processor macro or enum instead of directly using "2".  If there are many places where this style of freeing happens, a function to abstract this operation - eg "free_tag_value(tag_value, tag_allocated_flags)" - might provide extra narrative too.